diffrax/integrate.py

```diff
@@ -195,6 +203,168 @@ def _apply_history(x):
+                    direction=1,
+                )
+                flat_delays = jtu.tree_leaves(delays.delays)
+                _gs = []
+                for delay in flat_delays:
+                    g = (
+                        lambda t, delay=delay: t
+                        - delay(t, dense_discont.evaluate(t), args)
+                        - state.discontinuities
+                    )
+                    _gs.append(g)
+
+                def _find_discontinuity():
+                    # Start by doing a cheap bisection search to reduce
+                    # over the stored-discontinuity dimension.
+
+                    def _cond_fun(_val):
+                        _, _, _pred, _, _ = _val
+                        return _pred
+
+                    def _body_fun(_val):
+                        _ta, _tb, _, _step, max_step = _val
+                        _step = _step + 1
+                        _tmid = _ta + 0.5 * (_tb - _ta)
+                        _gas = jnp.stack([jnp.sign(g(_ta)) for g in _gs])
+                        _gmids = jnp.stack([jnp.sign(g(_tmid)) for g in _gs])
+                        _gbs = jnp.stack([jnp.sign(g(_tb)) for g in _gs])
+                        _any_left = jnp.any(_gas != _gmids)
+                        _next_ta = jnp.where(_any_left, _ta, _tmid)
+                        _next_tb = jnp.where(_any_left, _tmid, _tb)
+                        _pred = (
+                            jnp.any(jnp.sum(_gas != _gbs, axis=1) > 1) | _step
+                            > max_step
+                        )
+                        return _next_ta, _next_tb, _pred, _step, max_step
+
+                    _init_val = (sub_tprev, sub_tnext, True, 0, 400)
+                    _final_val = lax.while_loop(_cond_fun, _body_fun, _init_val)
+                    _ta, _tb, _, _, _ = _final_val
+
+                    # Then do a more expensive Newton search
+                    # to find the first discontinuity.
+                    _discont_solver = NewtonNonlinearSolver(
+                        rtol=delays.rtol, atol=delays.atol
+                    )
+                    _disconts = []
+                    for g, delay in zip(_gs, flat_delays):
+                        changed_sign = jnp.sign(g(_ta)) != jnp.sign(g(_tb))
+                        if changed_sign.all() is False:
+                            pass
+                        else:
+                            _i = jnp.argmax(changed_sign)
+                            _d = state.discontinuities[_i]
+                            _h = (
+                                lambda t, args, delay=delay, _d=_d: t
+                                - delay(t, dense_discont.evaluate(t), args)
+                                - _d
+                            )
+                            _discont = _discont_solver(_h, _tb, args).root
+                            _disconts.append(_discont)
+                    if _discont.size == 0:
+                        return jnp.inf
```

Comment:
I don't think we can ever hit this branch?
(And indeed, as currently written it would crash, since it doesn't also return a bool.)

Reply:
Yup we never would hit this, no problem should be encountered since `NewtonNonlinearSolver` return `NaNs` if it fails right ?